src/ext/php7/engine_hooks.c

```diff
@@ -479,80 +448,362 @@ static int _dd_opcode_default_dispatch(zend_execute_data *execute_data) {
+        return _prev_icall_handler ? _prev_icall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    ddtrace_dispatch_t *dispatch = NULL;
+    if (!EX(call)->func || !_dd_should_trace_call(EX(call), EX(call)->func, &dispatch)) {
+        return _prev_icall_handler ? _prev_icall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    if (dispatch->options & DDTRACE_DISPATCH_INNERHOOK) {
+        return _dd_legacy_fcall_helper(execute_data, dispatch);
+    }
+    _dd_fcall_helper(execute_data, dispatch);
+    return _prev_icall_handler ? _prev_icall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+}
+
+static int _dd_do_ucall_handler(zend_execute_data *execute_data) {
+    // Assume that a neighboring extension could have incremented the opline before forwarding the handler
+    if (ZEND_DO_UCALL != EX(opline)->opcode) {
+        return _prev_ucall_handler ? _prev_ucall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    ddtrace_dispatch_t *dispatch = NULL;
+    if (!EX(call)->func || !_dd_should_trace_call(EX(call), EX(call)->func, &dispatch)) {
+        return _prev_ucall_handler ? _prev_ucall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    if (dispatch->options & DDTRACE_DISPATCH_INNERHOOK) {
+        return _dd_legacy_fcall_helper(execute_data, dispatch);
+    }
+    _dd_fcall_helper(execute_data, dispatch);
+    return _prev_ucall_handler ? _prev_ucall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+}
+
+static int _dd_do_fcall_handler(zend_execute_data *execute_data) {
+    // Assume that a neighboring extension could have incremented the opline before forwarding the handler
+    if (ZEND_DO_FCALL != EX(opline)->opcode) {
+        return _prev_fcall_handler ? _prev_fcall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    ddtrace_dispatch_t *dispatch = NULL;
+    if (!EX(call)->func || !_dd_should_trace_call(EX(call), EX(call)->func, &dispatch)) {
+        return _prev_fcall_handler ? _prev_fcall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    if (dispatch->options & DDTRACE_DISPATCH_INNERHOOK) {
+        return _dd_legacy_fcall_helper(execute_data, dispatch);
+    }
+    _dd_fcall_helper(execute_data, dispatch);
+    return _prev_fcall_handler ? _prev_fcall_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+}
+
+static int _dd_do_fcall_by_name_handler(zend_execute_data *execute_data) {
+    // Assume that a neighboring extension could have incremented the opline before forwarding the handler
+    if (ZEND_DO_FCALL_BY_NAME != EX(opline)->opcode) {
+        return _prev_fcall_by_name_handler ? _prev_fcall_by_name_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
+    ddtrace_dispatch_t *dispatch = NULL;
+    if (!EX(call)->func || !_dd_should_trace_call(EX(call), EX(call)->func, &dispatch)) {
+        return _prev_fcall_by_name_handler ? _prev_fcall_by_name_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
+    }
```

Comment:
Probably need to clang-format this one.
```suggestion
    ddtrace_dispatch_t *dispatch = NULL;
    if (ZEND_DO_FCALL_BY_NAME != EX(opline)->opcode || !EX(call)->func || !_dd_should_trace_call(EX(call), EX(call)->func, &dispatch)) {
        return _prev_fcall_by_name_handler ? _prev_fcall_by_name_handler(execute_data) : ZEND_USER_OPCODE_DISPATCH;
    }
```